components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/OpenAPIUtils.java

```diff
@@ -70,35 +70,68 @@ public static String getResourceAuthenticationScheme(OpenAPI openAPI, MessageCon
     public static String getScopesOfResource(OpenAPI openAPI, MessageContext synCtx) {
         Map<String, Object> vendorExtensions = getPathItemExtensions(synCtx, openAPI);
         if (vendorExtensions != null) {
-            return (String) vendorExtensions.get(APIConstants.SWAGGER_X_SCOPE);
+            String resourceScope = (String) vendorExtensions.get(APIConstants.SWAGGER_X_SCOPE);
+            if (resourceScope == null) {
+                // If x-scope not found in swagger, check for the scopes in security
+                ArrayList<String> securityScopes = getPathItemSecurityScopes(synCtx, openAPI);
+                if (securityScopes == null || securityScopes.isEmpty()) {
+                    return null;
+                } else {
+                    return securityScopes.get(0);
+                }
+            } else {
+                return resourceScope;
+            }
         }
         return null;
     }
 
     /**
-     * Return the scope-role mapping of an API
+     * Return the roles of a given scope attached to a resource using the API swagger.
      *
      * @param openAPI OpenAPI of the API
      * @param synCtx  The message containing resource request
-     * @return the scope-role mapping
+     * @param resourceScope  The scope of the resource
+     * @return the roles of the scope in the comma separated format
      */
-    public static ArrayList<LinkedHashMap> getScopeToRoleMappingOfApi(OpenAPI openAPI, MessageContext synCtx) {
-        Map<String, Object> vendorExtensions = getPathItemExtensions(synCtx, openAPI);
+    public static String getRolesOfScope(OpenAPI openAPI, MessageContext synCtx, String resourceScope) {
+        String resourceRoles = null;
 
+        Map<String, Object> vendorExtensions = getPathItemExtensions(synCtx, openAPI);
         if (vendorExtensions != null) {
-            String resourceScope = getScopesOfResource(openAPI, synCtx);
             if (StringUtils.isNotBlank(resourceScope)) {
-                LinkedHashMap swaggerWSO2Security = (LinkedHashMap) openAPI.getExtensions()
-                        .get(APIConstants.SWAGGER_X_WSO2_SECURITY);
-                if (swaggerWSO2Security != null) {
-                    LinkedHashMap swaggerObjectAPIM = (LinkedHashMap) swaggerWSO2Security
-                            .get(APIConstants.SWAGGER_OBJECT_NAME_APIM);
-                    if (swaggerObjectAPIM != null) {
-                        return (ArrayList<LinkedHashMap>) swaggerObjectAPIM.get(APIConstants.SWAGGER_X_WSO2_SCOPES);
+                if (openAPI.getExtensions() != null) {
+                    LinkedHashMap swaggerWSO2Security = (LinkedHashMap) openAPI.getExtensions()
+                            .get(APIConstants.SWAGGER_X_WSO2_SECURITY);
```

Comment:
I think you need to do a Null check for openAPI.getExtensions().get(APIConstants.SWAGGER_X_WSO2_SECURITY) as well because there could be a NPE when casting to LinkedHashMap when there is no when get(APIConstants.SWAGGER_X_WSO2_SECURITY) becomes null. We could add the null check to the if condition

Reply:
done

Reply:
done